hosted-cluster-config-operator/controllers/resources/resources.go

```diff
@@ -104,8 +157,179 @@ func (r *reconciler) reconcile(ctx context.Context) error {
+		return nil
+	}); err != nil {
+		errs = append(errs, fmt.Errorf("failed to reconcile dns config: %w", err))
+	}
+
+	return errors.NewAggregate(errs)
+}
+
+func (r *reconciler) reconcileNamespaces(ctx context.Context) error {
+	namespaceManifests := []struct {
+		manifest  func() *corev1.Namespace
+		reconcile func(*corev1.Namespace) error
+	}{
+		{manifest: manifests.NamespaceOpenShiftAPIServer},
+		{manifest: manifests.NamespaceOpenShiftControllerManager},
+		{manifest: manifests.NamespaceKubeAPIServer, reconcile: namespaces.ReconcileKubeAPIServerNamespace},
+		{manifest: manifests.NamespaceKubeControllerManager},
+		{manifest: manifests.NamespaceKubeScheduler},
+		{manifest: manifests.NamespaceEtcd},
+		{manifest: manifests.NamespaceIngress, reconcile: namespaces.ReconcileOpenShiftIngressNamespace},
+		{manifest: manifests.NamespaceAuthentication},
+	}
+
+	var errs []error
+	for _, m := range namespaceManifests {
+		ns := m.manifest()
+		if _, err := r.CreateOrUpdate(ctx, r.client, ns, func() error {
+			if m.reconcile != nil {
+				return m.reconcile(ns)
+			}
+			return nil
+		}); err != nil {
+			errs = append(errs, fmt.Errorf("failed to reconcile namespace %s: %w", ns.Name, err))
+		}
+	}
+
+	return errors.NewAggregate(errs)
+}
+
+func (r *reconciler) reconcileRBAC(ctx context.Context) error {
+	roles := []struct {
+		manifest  func() *rbacv1.ClusterRole
+		reconcile func(*rbacv1.ClusterRole) error
+	}{
+		{manifest: manifests.CSRApproverClusterRole, reconcile: rbac.ReconcileCSRApproverClusterRole},
+		{manifest: manifests.IngressToRouteControllerClusterRole, reconcile: rbac.ReconcileIngressToRouteControllerClusterRole},
+		{manifest: manifests.NamespaceSecurityAllocationControllerClusterRole, reconcile: rbac.ReconcileNamespaceSecurityAllocationControllerClusterRole},
+	}
+
+	roleBindings := []struct {
+		manifest  func() *rbacv1.ClusterRoleBinding
+		reconcile func(*rbacv1.ClusterRoleBinding) error
+	}{
+		{manifest: manifests.CSRApproverClusterRoleBinding, reconcile: rbac.ReconcileCSRApproverClusterRoleBinding},
+		{manifest: manifests.IngressToRouteControllerClusterRoleBinding, reconcile: rbac.ReconcileIngressToRouteControllerClusterRoleBinding},
+		{manifest: manifests.NamespaceSecurityAllocationControllerClusterRoleBinding, reconcile: rbac.ReconcileNamespaceSecurityAllocationControllerClusterRoleBinding},
+		{manifest: manifests.NodeBootstrapperClusterRoleBinding, reconcile: rbac.ReconcileNodeBootstrapperClusterRoleBinding},
+		{manifest: manifests.CSRRenewalClusterRoleBinding, reconcile: rbac.ReconcileCSRRenewalClusterRoleBinding},
+		{manifest: manifests.ServiceAccountIssuerDiscoveryClusterRoleBinding, reconcile: rbac.ReconcileServiceAccountIssuerDiscoveryClusterRoleBinding},
+	}
+
+	var errs []error
+	for _, m := range roles {
+		role := m.manifest()
+		if _, err := r.CreateOrUpdate(ctx, r.client, role, func() error {
+			if m.reconcile != nil {
```

Comment:
Reconcile is always non-nil and I think if it was nil, the resulting role or rolebinding would be invalid so lets remove this check?